docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,220 @@
+
+```proto
+service Rules {
+  /// Rules has info for all rules.
+  rpc Rules(RulesRequest) returns (stream RulesResponse);
+}
+```
+
+Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly, Thanos Ruler constructs the response based in its local state.
+
+Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:
+
+* Deduplication happens first the rule group level. The identifier is the group name.
+* Then, per group name deduplication happens on the rule level, where:
+
+1. the rule type (recording rule vs. alerting rule)
+2. the rule name
+3. the rule label names
+4. the rule expression `expr` field
+5. the alerting rule `for` field
+
+are being used as the deduplication identifier. Disjunct entries are simply merged by adding them to the result set.
+
+Thanos query presents the result of the fan-out on a `/api/v1/rules` endpoint which is compatible with the Prometheus' rules API endpoint.
+Additionally Thanos Querier gains a new `--rule.replica-label` command line argument which falls back to `--query.replica-label` if unset. The replica labels refer to the same labels as specified in the `external_labels` section of Prometheus and the `--label` command line argument of Thanos Ruler.
+
+#### Examples
+
+A stream of alerting rules is merged from different Thanos Ruler API instances.
+These may be remote Thanos Ruler, Prometheus Sidecar, or even Thanos Querier instances.
+The merging Thanos Querier has `--rule.replica-label=replica`
+
+Scenario 1:
+
+As specified, the rule type and then rule name is used for deduplication.
+
+Given the following stream of incoming rule groups and containing recording/alerting rules:
+```text
+group: a
+   recording:<name:"a1" last_evaluation:<seconds:1 > >
+   alert:    <name:"r1" last_evaluation:<seconds:1 > >
+group: b
+   recording:<name:"r1" last_evaluation:<seconds:1 > >
+group: a
+   recording:<name:"r2" last_evaluation:<seconds:1 > >
+```
+
+The output becomes:
+```text
+group: a
+   alert:    <name:"a1" last_evaluation:<seconds:1 > >
+   recording:<name:"r1" last_evaluation:<seconds:1 > >
+   recording:<name:"r2" last_evaluation:<seconds:1 > >
+group: b
+   recording:<name:"r1" last_evaluation:<seconds:1 > >
+```
+
+Note in the example above how the recording rule `r1` is not deduplicated as it is contained in two different groups.
+
+Scenario 2:
+
+The next level of deduplication is governed by the label/value set of the underlying recording/alerting rule while respecting the replica label.
+For a given conflict, the youngest is preferred. For alerting rules, the youngest firing rule is preferred.
```

Comment:
You mean this, right, as in the example that immediately follows this?

```suggestion
For a given recording rules conflict, the oldest is preferred. For alerting rules, the youngest firing rule is preferred.
```

Reply:
2 seconds is meant to be younger than 1 second :-) I meant youngest, yes. I'll change the examples to use full dates instead.